### PR TITLE
runtime: check AVR quote for debug bit

### PR DESCRIPTION
**Pull request by matevz**
PR for https://github.com/oasislabs/ekiden/issues/2181

This PR:
- adds `OASIS_UNSAFE_ALLOW_DEBUG_ENCLAVES` env variable at compile time which checks AVR attributes Debug flag in rust,
- adds `ias.debug.allow_debug_enclaves` runtime switch for worker which checks AVR attributes Debug flag in go,
- if debug enclaves are not allowed, they are dropped; similarly, if debug enclaves are allowed, production enclaves are dropped.

### CODE REVIEW

File: go/common/sgx/ias/quote.go [hunk 1]
```diff
@@ -146,6 +165,14 @@ func (q *Quote) Verify() error {
 	if mrSignerBlacklist[q.Report.MRSIGNER] {
 		return fmt.Errorf("ias/quote: blacklisted MRSIGNER")
 	}
+
+	// Drop enclaves compiled in debug mode.
+	if os.Getenv("OASIS_UNSAFE_ALLOW_DEBUG_ENLCAVES") == "" {
```
[the remark this answers is not in the document]
Comment (kostko) [reply go/common/sgx/ias/quote.go]: Also I'm not sure if we want to set this via an env var on the Go side? Shouldn't it be the same as "skip AVR verify" (e.g., a special flag)?

File: go/common/node/node.go [hunk 2]
```diff
@@ -460,6 +460,11 @@ func (c *CapabilityTEE) Verify(ts time.Time) error {
 			return err
 		}
 
+		// Verify the quote.
+		if err = q.Verify(); err != nil {
+			return err
+		}
```
Comment (pro-wh) [inline go/common/node/node.go]: why didn't we do this before? will adding this break things?

File: go/common/sgx/ias/ias.go [hunk 3]
```diff
@@ -101,6 +102,13 @@ func (e *httpEndpoint) VerifyEvidence(ctx context.Context, quoteBinary, pseManif
 		return nil, nil, nil, fmt.Errorf("ias: invalid nonce length")
 	}
 
+	// Drop enclaves compiled in debug mode.
+	if os.Getenv("OASIS_UNSAFE_ALLOW_DEBUG_ENCLAVES") == "" {
+		if quote.Report.Attributes.GetFlag(AttributeDebug) {
+			return nil, nil, nil, fmt.Errorf("ias: dropped enclave running in debug mode")
+		}
+	}
```
Comment (pro-wh) [inline go/common/sgx/ias/ias.go]: would it be okay to upgrade this to a `quite.Verify()`?

File: go/common/sgx/ias/quote.go [hunk 4]
```diff
@@ -85,11 +86,29 @@ func (b *Body) MarshalBinary() ([]byte, error) {
 	return bBin, nil
 }
 
+// Attributes is a SGX enclave attributes value.
+type Attributes [16]byte
+type AttributeFlag uint8
```
Comment (pro-wh) [inline go/common/sgx/ias/quote.go]: why are we using uint8? other materials are modeling it as uint64

File: runtime/src/common/sgx/avr.rs [hunk 5]
```diff
@@ -316,6 +324,22 @@ pub fn verify(avr: &AVR) -> Fallible<AuthenticatedAVR> {
         _ => return Err(AVRError::MalformedQuote.into()),
     };
 
+    #[cfg(target_env = "sgx")]
+    {
+        if option_env!("OASIS_UNSAFE_ALLOW_DEBUG_ENCLAVES").is_none() {
+            let af = match AttributesFlags::from_bits(u64::from_le_bytes(
+                quote_body.report_body.attributes[0..8].try_into()?,
```
Comment (pro-wh) [inline runtime/src/common/sgx/avr.rs]: sorry for the scope creep: would it be possible to use sgx_isa to do the parsing of the whole report?

File: .buildkite/pipeline.yml [hunk 6]
```diff
@@ -39,6 +39,7 @@ docker_plugin_sgx_config: &docker_plugin_sgx_config
     - /dev/isgx
   environment:
     - "OASIS_TEE_HARDWARE=intel-sgx"
+    - "OASIS_UNSAFE_ALLOW_DEBUG_ENCLAVES=1"
```
Comment (pro-wh) [inline .buildkite/pipeline.yml]: why is this alone here? could this be the wrong place to put it?

File: go/common/node/node.go (hunk 2 above)
Comment (matevz) [reply go/common/node/node.go]: It seems to work fine.

File: .buildkite/pipeline.yml (hunk 6 above)
Comment (matevz) [reply .buildkite/pipeline.yml]: For go, `OASIS_UNSAFE_ALLOW_DEBUG_ENCLAVES` is a runtime parameter. We could use `-ldflags` switch to set a compile-time flag, but this seemed even worse.

File: go/common/sgx/ias/quote.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (kostko) [reply go/common/sgx/ias/quote.go]: As I mentioned above, this should not be configured via an environment variable on the Go side. It should be similar to how we handle "skip AVR verify".

File: runtime/src/common/sgx/avr.rs (hunk 5 above)
Comment (matevz) [reply runtime/src/common/sgx/avr.rs]: done

File: .buildkite/pipeline.yml (hunk 6 above)
Comment (pro-wh) [reply .buildkite/pipeline.yml]: ah, right. let's unify discussion on Jernej's comment about that

File: runtime/src/common/sgx/avr.rs [hunk 7]
```diff
@@ -15,8 +12,9 @@ use serde_json;
 use untrusted;
 use webpki;
 
-#[cfg(target_env = "sgx")]
-use sgx_isa::Report;
+use byteorder::{LittleEndian, ReadBytesExt};
+use sgx_isa::{AttributesFlags, Report};
+use std::io::{Cursor, Read, Seek, SeekFrom};
```
Comment (pro-wh) [inline runtime/src/common/sgx/avr.rs]: now that these aren't conditioned under `target_env`, we can put them into the std -> 3rd party dependencies -> local dependencies order

File: runtime/src/common/sgx/avr.rs (hunk 5 above)
Comment (pro-wh) [reply runtime/src/common/sgx/avr.rs]: thanks!

File: runtime/src/common/sgx/avr.rs [hunk 8]
```diff
@@ -314,44 +287,59 @@ pub fn verify(avr: &AVR) -> Fallible<AuthenticatedAVR> {
...
+    {
+        // Force-ratchet the clock forward, to at least the time in the AVR.
+        update_insecure_posix_time(timestamp);
+
+        // Return dummy AVR.
+        Ok(AuthenticatedAVR {
+            report_data: vec![],
+            identity: EnclaveIdentity {
+                mr_enclave: MrEnclave::from(vec![]),
+                mr_signer: MrSigner::from(vec![]),
```
Comment (pro-wh) [inline runtime/src/common/sgx/avr.rs]: was this desired? that we not do the usual checks and return a dummy AVR? I thought we use this in the gateway

File: go/common/sgx/ias/quote.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (matevz) [reply go/common/sgx/ias/quote.go]: Agree, I added `ias.debug.allow_debug_enclaves` go runtime flag now instead. 
`OASIS_UNSAFE_ALLOW_DEBUG_ENLCAVES` is used in compile-time rust.

File: runtime/src/common/sgx/avr.rs (hunk 8 above)
Comment (matevz) [reply runtime/src/common/sgx/avr.rs]: Yeah, this was a leftover of `Report` cleanup. SGX/non-SGX case can now be joined.

File: go/common/sgx/ias/avr.go [hunk 9]
```diff
@@ -20,6 +20,7 @@ import (
 const nonceMaxLen = 32
 
 var unsafeSkipVerify bool
+var unsafeAllowDebugEnclaves bool
```
Comment (Yawning) [inline go/common/sgx/ias/avr.go]: Use a var block.

```
var (
   foo bool
   bar uint32
)
```

File: go/common/sgx/ias/quote.go [hunk 10]
```diff
@@ -146,6 +164,14 @@ func (q *Quote) Verify() error {
 	if mrSignerBlacklist[q.Report.MRSIGNER] {
 		return fmt.Errorf("ias/quote: blacklisted MRSIGNER")
 	}
+
+	// Drop enclaves compiled in debug mode.
```
Comment (Yawning) [inline go/common/sgx/ias/quote.go]: Random thought.  Should the unsafe flag, mandate debug enclaves?  IMO it should be either production enclaves OR debug enclaves, not "set a flag and be allowed to mix both".

File: go/common/sgx/ias/avr_test.go [hunk 11]
```diff
@@ -14,6 +14,9 @@ func TestAVR(t *testing.T) {
 }
 
 func testAVRv2(t *testing.T) {
+	// TODO: Generate another testdata AVR with debug bit disabled.
+	unsafeAllowDebugEnclaves = true
```
Comment (Yawning) [inline go/common/sgx/ias/avr_test.go]: Unset this in a defer so that other tests that depend on verification being strict do the right thing.

File: go/common/sgx/ias/quote_test.go [hunk 12]
```diff
@@ -10,6 +10,9 @@ import (
 )
 
 func TestQuote(t *testing.T) {
+	// TODO: Generate another testdata AVR with debug bit disabled.
+	unsafeAllowDebugEnclaves = true
```
Comment (Yawning) [inline go/common/sgx/ias/quote_test.go]: Likewise, unset this in a defer.

File: go/common/sgx/ias/quote_test.go [hunk 13]
```diff
@@ -38,12 +41,8 @@ func TestQuote(t *testing.T) {
 		"CPUSVN",
 	)
 	require.EqualValues(t, 0x00000000, quote.Report.MiscSelect, "MISCSELECT")
-	require.Equal(
-		t,
-		"07000000000000000700000000000000",
-		hex.EncodeToString(quote.Report.Attributes[:]),
-		"ATTRIBUTES",
-	)
+	require.EqualValues(t, 0x0000000000000007, quote.Report.Attributes.Flags, "ATTRIBUTES.FLAGS")
```
Comment (Yawning) [inline go/common/sgx/ias/quote_test.go]: God I hope that we don't have XFRM and the flags transposed.

Comment (matevz) [reply go/common/sgx/ias/quote_test.go]: We don't ;)

File: go/common/sgx/ias/quote.go (hunk 10 above)
Comment (matevz) [reply go/common/sgx/ias/quote.go]: Hm.. but this will prevent us from doing any tests with the production enclave. Will this never be the case?

**Comment by Yawning** [comment]
So I had a thought about this, that might lead to more scope creep.  Given that we are doing #965, should allowing debug enclaves be consensus backed instead?

 1. We have a network wide consensus parameter that indicates if we allow debug enclaves or not.
 2. Each runtime's registry descriptor (`go/registry/runtime.go:Runtime`) gets a field that indicates if it is a debug enclave or not (Maybe move the exiting `TEEHardware` field into a struct or something?)
 3. On runtime registration (currently only in the genesis block, but the code is generic), if the newly registered runtime is a debug enclave, check to see if the consensus parameter (1) to allow such things is set, if not, reject the registration.
 4. Nodes that indicate runtime support MUST have the debug status match what is in the runtime descriptor (2).

For now `1` can be faked with a flag.  The other stuff can be implemented in a way that once we have consensus backed params, the changes should be minor (especially since the IAS proxy can query a node or interpret a genesis document).

**Comment by kostko** [comment]
> should allowing debug enclaves be consensus backed instead?

I think this (and the proposed design) is a good idea.

So the runtimes would still need to respect the compile-time value of the parameter when validating AVRs as they currently can't trust consensus (that would require compiling-in the genesis block anyway which seems equivalent to just reading the compile-time env variable).

**Comment by Yawning** [comment]
> So the runtimes would still need to respect the compile-time value of the parameter when validating AVRs as they currently can't trust consensus (that would require compiling-in the genesis block anyway which seems equivalent to just reading the compile-time env variable).

Yeah, this is just to prevent misconfigured nodes from doing the wrong thing, by forcing the policy to be uniform (MRENCLAVE will mismatch if the enclave is built with the wrong flags, so we're ok there.).

File: go/common/sgx/ias/quote.go (hunk 10 above)
Comment (matevz) [reply go/common/sgx/ias/quote.go]: Fixed.

**Comment by matevz** [comment]
I opened a separate issue https://github.com/oasislabs/oasis-core/issues/2285 for having on-chain debug enclaves flag. This PR addresses the CLI part only.

File: go/common/node/node.go (hunk 2 above)
Comment (kostko) [reply go/common/node/node.go]: This is actually not needed as `avrBundle.Open` already calls `validate` which calls `Verify` on the quote. So this was already being verified in `Open` and can be removed.

File: go/common/sgx/ias/ias.go [hunk 14]
```diff
@@ -101,6 +101,11 @@ func (e *httpEndpoint) VerifyEvidence(ctx context.Context, quoteBinary, pseManif
 		return nil, nil, nil, fmt.Errorf("ias: invalid nonce length")
 	}
 
+	// Drop enclaves compiled in debug mode.
```
Comment (kostko) [inline go/common/sgx/ias/ias.go]: This comment seems out of place? This verifies the quote.

File: go/common/sgx/ias/quote.go [hunk 15]
```diff
@@ -146,6 +174,19 @@ func (q *Quote) Verify() error {
 	if mrSignerBlacklist[q.Report.MRSIGNER] {
 		return fmt.Errorf("ias/quote: blacklisted MRSIGNER")
 	}
+
+	if !unsafeAllowDebugEnclaves {
+		// Drop debug enclaves, if we are in production mode.
```
Comment (kostko) [inline go/common/sgx/ias/quote.go]: Maybe instead of "drop" use the term "disallow" (everywhere, also in error message below).

File: go/common/sgx/ias/quote_test.go [hunk 16]
```diff
@@ -10,6 +10,10 @@ import (
 )
 
 func TestQuote(t *testing.T) {
+	// TODO: Generate another testdata AVR with debug bit disabled.
```
Comment (kostko) [inline go/common/sgx/ias/quote_test.go]: Can we get one? cc @pro-wh

File: go/ias/ias.go [hunk 17]
```diff
@@ -209,18 +210,24 @@ func New(identity *identity.Identity) (*IAS, error) {
...
 	return s, nil
 }
 
 func init() {
-	Flags.String(cfgProxyAddress, "", "IAS proxy address")
-	Flags.String(cfgTLSCertFile, "", "IAS proxy TLS certificate")
-	Flags.Bool(cfgDebugSkipVerify, false, "skip IAS AVR signature verification (UNSAFE)")
+	Flags.String(CfgProxyAddress, "", "IAS proxy address")
+	Flags.String(CfgTLSCertFile, "", "IAS proxy TLS certificate")
+	Flags.Bool(CfgDebugSkipVerify, false, "skip IAS AVR signature verification (UNSAFE)")
```
Comment (kostko) [inline go/ias/ias.go]: I would hide the unsafe flags (e.g., using `Flags.MarkHidden`).

File: go/registry/api/api.go [hunk 18]
```diff
@@ -80,6 +80,10 @@ var (
 	// register a runtime with bad Capabilities.TEE.Hardware.
 	ErrBadCapabilitiesTEEHardware = errors.New("registry: bad capabilities.TEE.Hardware")
 
+	// ErrDebugEnclave is the error returned when a node tries to register
+	// runtimes with enclaves running in debug mode.
+	ErrDebugEnclave = errors.New("registry: enclave running in debug mode not allowed")
```
Comment (kostko) [inline go/registry/api/api.go]: Is the same error used if the node tries to register a production enclave but the network is running in debug mode? In this case the error description (and doc comment) should probably be updated.

File: runtime/src/common/sgx/avr.rs [hunk 19]
```diff
@@ -316,14 +298,36 @@ pub fn verify(avr: &AVR) -> Fallible<AuthenticatedAVR> {
         _ => return Err(AVRError::MalformedQuote.into()),
     };
 
+    if option_env!("OASIS_UNSAFE_ALLOW_DEBUG_ENCLAVES").is_none() {
+        // Drop debug enclaves, if we are in production mode.
```
Comment (kostko) [inline runtime/src/common/sgx/avr.rs]: Same as above -- disallow instead of drop.

File: runtime/src/common/sgx/avr.rs [hunk 20]
```diff
@@ -316,14 +298,36 @@ pub fn verify(avr: &AVR) -> Fallible<AuthenticatedAVR> {
...
+        if quote_body
```
Comment (kostko) [inline runtime/src/common/sgx/avr.rs]: This could be simplified by something like:
```rust
let is_debug = quote_body.report_body...contains(AttributeFlags::DEBUG);
let allow_debug = option_env!("...").is_none();
if is_debug != allow_debug { /* error */ }
```

File: go/common/sgx/ias/quote_test.go (hunk 16 above)
Comment (matevz) [reply go/common/sgx/ias/quote_test.go]: Does this mean AVR needs to be signed with production key? (i.e. can you turn off debug bit and sign it with testing key?)

Comment (matevz) [reply go/common/sgx/ias/quote_test.go]: We cannot obtain AVR signed with the production key at the moment. Also confirmed that you cannot generate AVR with debug bit disabled without the production key. I am resolving this for now.
